src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -451,198 +480,168 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
         return _connection.State();
     }
 
-    winrt::fire_and_forget TermControl::SwapChainChanged()
+    winrt::fire_and_forget TermControl::RenderEngineSwapChainChanged()
     {
-        if (!_initializedTerminal)
-        {
-            return;
+        { // lock scope
+            auto terminalLock = _terminal->LockForReading();
+            if (!_initializedTerminal)
+            {
+                return;
+            }
         }
 
         auto chain = _renderEngine->GetSwapChain();
         auto weakThis{ get_weak() };
 
         co_await winrt::resume_foreground(Dispatcher());
 
-        // If 'weakThis' is locked, then we can safely work with 'this'
         if (auto control{ weakThis.get() })
         {
-            if (_terminal)
-            {
-                auto lock = _terminal->LockForWriting();
-                auto nativePanel = SwapChainPanel().as<ISwapChainPanelNative>();
-                nativePanel->SetSwapChain(chain.Get());
-            }
+            auto terminalLock = _terminal->LockForWriting();
+
+            _AttachDxgiSwapChainToXaml(chain.Get());
         }
     }
 
-    winrt::fire_and_forget TermControl::_SwapChainRoutine()
+    void TermControl::_AttachDxgiSwapChainToXaml(IDXGISwapChain1* swapChain)
     {
-        auto chain = _renderEngine->GetSwapChain();
-        auto weakThis{ get_weak() };
-
-        co_await winrt::resume_foreground(Dispatcher());
-
-        if (auto control{ weakThis.get() })
-        {
-            if (_terminal)
-            {
-                auto lock = _terminal->LockForWriting();
-                auto nativePanel = SwapChainPanel().as<ISwapChainPanelNative>();
-                nativePanel->SetSwapChain(chain.Get());
-            }
-        }
+        auto nativePanel = SwapChainPanel().as<ISwapChainPanelNative>();
+        nativePanel->SetSwapChain(swapChain);
     }
 
     bool TermControl::_InitializeTerminal()
     {
-        if (_initializedTerminal)
-        {
-            return false;
-        }
-
-        const auto windowWidth = SwapChainPanel().ActualWidth(); // Width() and Height() are NaN?
-        const auto windowHeight = SwapChainPanel().ActualHeight();
-
-        if (windowWidth == 0 || windowHeight == 0)
-        {
-            return false;
-        }
+        { // scope for terminalLock

```

Comment:
Would it make more sense to just `unlock` the lock instead of introducing another scope? idk

Reply:
Generally I prefer scopes for these, but it is largely stylistic